File: backend/app/services/gxe_analysis.py
"""
G×E Interaction Analysis Service
Genotype-by-Environment interaction analysis for plant breeding

Methods:
- AMMI (Additive Main effects and Multiplicative Interaction)
- GGE Biplot
- Finlay-Wilkinson Regression

Uses NumPy/SciPy for calculations (Fortran bindings available for production)
"""

import numpy as np
from typing import Optional, List, Dict, Any, Tuple
from dataclasses import dataclass, field
from enum import Enum
import logging

logger = logging.getLogger(__name__)


class GxEMethod(str, Enum):
    """Available G×E analysis methods"""
    AMMI = "ammi"
    GGE = "gge"
    FINLAY_WILKINSON = "finlay_wilkinson"


class GGEScaling(int, Enum):
    """GGE biplot scaling options"""
    SYMMETRIC = 0
    GENOTYPE_FOCUSED = 1
    ENVIRONMENT_FOCUSED = 2


@dataclass
class AMMIResult:
    """AMMI analysis results"""
    grand_mean: float
    genotype_effects: np.ndarray
    environment_effects: np.ndarray
    genotype_scores: np.ndarray  # IPCA scores
    environment_scores: np.ndarray
    eigenvalues: np.ndarray
    variance_explained: np.ndarray
    genotype_names: List[str]
    environment_names: List[str]
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "grand_mean": float(self.grand_mean),
            "genotype_effects": self.genotype_effects.tolist(),
            "environment_effects": self.environment_effects.tolist(),
            "genotype_scores": self.genotype_scores.tolist(),
            "environment_scores": self.environment_scores.tolist(),
            "eigenvalues": self.eigenvalues.tolist(),
            "variance_explained": self.variance_explained.tolist(),
            "genotype_names": self.genotype_names,
            "environment_names": self.environment_names,
        }


@dataclass
class GGEResult:
    """GGE biplot results"""
    genotype_scores: np.ndarray
    environment_scores: np.ndarray
    eigenvalues: np.ndarray
    variance_explained: np.ndarray
    genotype_names: List[str]
    environment_names: List[str]
    scaling: str
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "genotype_scores": self.genotype_scores.tolist(),
            "environment_scores": self.environment_scores.tolist(),
            "eigenvalues": self.eigenvalues.tolist(),
            "variance_explained": self.variance_explained.tolist(),
            "genotype_names": self.genotype_names,
            "environment_names": self.environment_names,
            "scaling": self.scaling,
        }


@dataclass 
class FinlayWilkinsonResult:
    """Finlay-Wilkinson regression results"""
    genotype_means: np.ndarray
    slopes: np.ndarray  # Stability parameter (b)
    r_squared: np.ndarray
    environment_index: np.ndarray
    genotype_names: List[str]
    environment_names: List[str]
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "genotype_means": self.genotype_means.tolist(),
            "slopes": self.slopes.tolist(),
            "r_squared": self.r_squared.tolist(),
            "environment_index": self.environment_index.tolist(),
            "genotype_names": self.genotype_names,
            "environment_names": self.environment_names,
            "stability_classification": self._classify_stability(),
        }
    
    def _classify_stability(self) -> List[Dict[str, Any]]:
        """Classify genotypes by stability"""
        results = []
        mean_yield = np.mean(self.genotype_means)
        
        for i, name in enumerate(self.genotype_names):
            slope = self.slopes[i]
            mean = self.genotype_means[i]
            
            # Classification based on Finlay-Wilkinson interpretation
            if slope < 0.8:
                stability = "stable_low_responsive"
                desc = "Stable, below-average response to environment"
            elif slope > 1.2:
                stability = "unstable_high_responsive" 
                desc = "Unstable, above-average response to environment"
            else:
                stability = "average_responsive"
                desc = "Average response to environment"
            
            if mean > mean_yield:
                performance = "high_yielding"
            else:
                performance = "low_yielding"
            
            results.append({
                "genotype": name,
                "mean": float(mean),
                "slope": float(slope),
                "r_squared": float(self.r_squared[i]),
                "stability": stability,
                "performance": performance,
                "description": desc,
            })
        
        return results


class GxEAnalysisService:
    """
    G×E Interaction Analysis Service
    
    Provides AMMI, GGE biplot, and Finlay-Wilkinson regression
    for analyzing genotype-by-environment interactions.
    """
    
    def ammi_analysis(
        self,
        yield_matrix: np.ndarray,
        genotype_names: List[str],
        environment_names: List[str],
        n_components: int = 2
    ) -> AMMIResult:
        """
        AMMI (Additive Main effects and Multiplicative Interaction) Analysis
        
        Model: Y_ij = μ + G_i + E_j + Σ λ_k α_ik γ_jk + ε_ij
        
        Args:
            yield_matrix: Yield data (genotypes × environments)
            genotype_names: Names of genotypes
            environment_names: Names of environments
            n_components: Number of IPCA components to retain
            
        Returns:
            AMMIResult with effects, scores, and eigenvalues
        """
        g, e = yield_matrix.shape
        n_components = min(n_components, min(g, e) - 1)
        
        # Grand mean
        grand_mean = np.mean(yield_matrix)
        
        # Genotype effects (row means - grand mean)
        genotype_effects = np.mean(yield_matrix, axis=1) - grand_mean
        
        # Environment effects (column means - grand mean)
        environment_effects = np.mean(yield_matrix, axis=0) - grand_mean
        
        # GE interaction matrix
        GE = yield_matrix - grand_mean - genotype_effects[:, np.newaxis] - environment_effects
        
        # SVD of interaction matrix
        U, S, Vt = np.linalg.svd(GE, full_matrices=False)
        
        # IPCA scores (scaled by sqrt of singular values)
        genotype_scores = U[:, :n_components] * np.sqrt(S[:n_components])
        environment_scores = Vt[:n_components, :].T * np.sqrt(S[:n_components])
        
        # Eigenvalues and variance explained
        eigenvalues = S[:n_components] ** 2
        total_ss = np.sum(GE ** 2)
        variance_explained = eigenvalues / total_ss * 100 if total_ss > 0 else eigenvalues * 0
        
        return AMMIResult(
            grand_mean=grand_mean,
            genotype_effects=genotype_effects,
            environment_effects=environment_effects,
            genotype_scores=genotype_scores,
            environment_scores=environment_scores,
            eigenvalues=eigenvalues,
            variance_explained=variance_explained,
            genotype_names=genotype_names,
            environment_names=environment_names,
        )
    
    def gge_biplot(
        self,
        yield_matrix: np.ndarray,
        genotype_names: List[str],
        environment_names: List[str],
        n_components: int = 2,
        scaling: GGEScaling = GGEScaling.SYMMETRIC
    ) -> GGEResult:
        """
        GGE (Genotype + Genotype-by-Environment) Biplot Analysis
        
        Model: Y_ij - μ - E_j = Σ λ_k α_ik γ_jk
        Environment-centered, combines G + GE effects
        
        Args:
            yield_matrix: Yield data (genotypes × environments)
            genotype_names: Names of genotypes
            environment_names: Names of environments
            n_components: Number of PCs to retain
            scaling: Scaling method for biplot
            
        Returns:
            GGEResult with scores and eigenvalues
        """
        g, e = yield_matrix.shape
        n_components = min(n_components, min(g, e) - 1)
        
        # Environment-center the data (remove environment means)
        env_means = np.mean(yield_matrix, axis=0)
        Y_centered = yield_matrix - env_means
        
        # SVD
        U, S, Vt = np.linalg.svd(Y_centered, full_matrices=False)
        
        # Apply scaling
        if scaling == GGEScaling.SYMMETRIC:
            scale_g = np.sqrt(S[:n_components])
            scale_e = np.sqrt(S[:n_components])
        elif scaling == GGEScaling.GENOTYPE_FOCUSED:
            scale_g = S[:n_components]
            scale_e = np.ones(n_components)
        else:  # ENVIRONMENT_FOCUSED
            scale_g = np.ones(n_components)
            scale_e = S[:n_components]
        
        genotype_scores = U[:, :n_components] * scale_g
        environment_scores = Vt[:n_components, :].T * scale_e
        
        # Variance explained
        eigenvalues = S[:n_components] ** 2
        total_ss = np.sum(Y_centered ** 2)
        variance_explained = eigenvalues / total_ss * 100 if total_ss > 0 else eigenvalues * 0
        
        scaling_names = {0: "symmetric", 1: "genotype_focused", 2: "environment_focused"}
        
        return GGEResult(
            genotype_scores=genotype_scores,
            environment_scores=environment_scores,
            eigenvalues=eigenvalues,
            variance_explained=variance_explained,
            genotype_names=genotype_names,
            environment_names=environment_names,
            scaling=scaling_names.get(scaling.value, "symmetric"),
        )
    
    def finlay_wilkinson(
        self,
        yield_matrix: np.ndarray,
        genotype_names: List[str],
        environment_names: List[str]
    ) -> FinlayWilkinsonResult:
        """
        Finlay-Wilkinson Regression Analysis
        
        Model: Y_ij = μ_i + β_i * E_j + ε_ij
        where E_j is the environment index (mean - grand mean)
        
        Interpretation:
        - β ≈ 1: Average stability
        - β < 1: Below-average response (stable in poor environments)
        - β > 1: Above-average response (responsive to good environments)
        
        Args:
            yield_matrix: Yield data (genotypes × environments)
            genotype_names: Names of genotypes
            environment_names: Names of environments
            
        Returns:
            FinlayWilkinsonResult with means, slopes, and R²
        """
        g, e = yield_matrix.shape
        
        # Grand mean and environment index
        grand_mean = np.mean(yield_matrix)
        env_means = np.mean(yield_matrix, axis=0)
        env_index = env_means - grand_mean
        
        # Genotype means
        genotype_means = np.mean(yield_matrix, axis=1)
        
        # Regression for each genotype
        slopes = np.zeros(g)
        r_squared = np.zeros(g)
        
        for i in range(g):
            y = yield_matrix[i, :]
            x = env_index
            
            # Simple linear regression
            x_mean = np.mean(x)
            y_mean = np.mean(y)
            
            ss_xy = np.sum((x - x_mean) * (y - y_mean))
            ss_xx = np.sum((x - x_mean) ** 2)
            ss_yy = np.sum((y - y_mean) ** 2)
            
            if ss_xx > 1e-10:
                slopes[i] = ss_xy / ss_xx
            else:
                slopes[i] = 1.0
            
            if ss_yy > 1e-10 and ss_xx > 1e-10:
                r_squared[i] = (ss_xy ** 2) / (ss_xx * ss_yy)
            else:
                r_squared[i] = 0.0
        
        return FinlayWilkinsonResult(
            genotype_means=genotype_means,
            slopes=slopes,
            r_squared=r_squared,
            environment_index=env_index,
            genotype_names=genotype_names,
            environment_names=environment_names,
        )
    
    def identify_mega_environments(
        self,
        gge_result: GGEResult,
        n_clusters: int = 3
    ) -> Dict[str, Any]:
        """
        Identify mega-environments from GGE biplot
        
        Uses environment scores to cluster similar environments.
        """
        from scipy.cluster.hierarchy import linkage, fcluster
        
        env_scores = gge_result.environment_scores
        
        # Hierarchical clustering
        Z = linkage(env_scores, method='ward')
        clusters = fcluster(Z, n_clusters, criterion='maxclust')
        
        mega_envs = {}
        for i, (name, cluster) in enumerate(zip(gge_result.environment_names, clusters)):
            cluster_name = f"ME{cluster}"
            if cluster_name not in mega_envs:
                mega_envs[cluster_name] = []
            mega_envs[cluster_name].append(name)
        
        return {
            "mega_environments": mega_envs,
            "cluster_assignments": dict(zip(gge_result.environment_names, clusters.tolist())),
        }
    
    def identify_winning_genotypes(
        self,
        gge_result: GGEResult
    ) -> Dict[str, Any]:
        """
        Identify winning genotypes for each environment
        
        Based on GGE biplot: genotype closest to environment vector wins.
        """
        g_scores = gge_result.genotype_scores
        e_scores = gge_result.environment_scores
        
        winners = {}
        for j, env_name in enumerate(gge_result.environment_names):
            env_vec = e_scores[j]
            
            # Project genotypes onto environment vector
            projections = np.dot(g_scores, env_vec) / (np.linalg.norm(env_vec) + 1e-10)
            winner_idx = np.argmax(projections)
            
            winners[env_name] = {
                "winner": gge_result.genotype_names[winner_idx],
                "projection": float(projections[winner_idx]),
                "rankings": [
                    {"genotype": gge_result.genotype_names[i], "score": float(projections[i])}
                    for i in np.argsort(projections)[::-1][:5]
                ]
            }
        
        return winners




    async def get_observation_matrix(
        self,
        db,  # AsyncSession
        study_db_ids: List[str],
        trait_db_id: str
    ) -> Dict[str, Any]:
        """
        Generate Yield Matrix from Database Observations
        
        Args:
            db: Database session
            study_db_ids: List of Study DB IDs to include (Environments)
            trait_db_id: Trait DB ID to analyze (Yield)
            
        Returns:
            Dict containing yield_matrix, genotype_names, environment_names
        """
        from sqlalchemy import select, func, cast, Float
        from app.models.phenotyping import Observation, ObservationUnit, ObservationVariable
        from app.models.core import Study
        from app.models.germplasm import Germplasm
        
        # 1. Fetch observations
        stmt = (
            select(
                Germplasm.germplasm_name,
                Study.study_name,
                func.avg(cast(Observation.value, Float)).label("mean_value")
            )
            .join(ObservationUnit, Observation.observation_unit_id == ObservationUnit.id)
            .join(Germplasm, Observation.germplasm_id == Germplasm.id)
            .join(Study, Observation.study_id == Study.id)
            .join(ObservationVariable, Observation.observation_variable_id == ObservationVariable.id) 
            .where(ObservationVariable.observation_variable_db_id == trait_db_id)
            .where(Study.study_db_id.in_(study_db_ids))
            .group_by(Germplasm.germplasm_name, Study.study_name)
        )
        
        result = await db.execute(stmt)
        rows = result.all()
        
        if not rows:
            return {
                "yield_matrix": [],
                "genotype_names": [],
                "environment_names": [],
                "error": "No observations found for these criteria"
            }
            
        # 2. Pivot Data
        # structured as {(genotype, environment): value}
        data_map = {(r.germplasm_name, r.study_name): r.mean_value for r in rows}
        
        # Get unique sorted lists
        genotypes = sorted(list(set(r.germplasm_name for r in rows)))
        environments = sorted(list(set(r.study_name for r in rows)))
        
        # Build Matrix
        matrix = []
        for g in genotypes:
            row = []
            for e in environments:
                val = data_map.get((g, e))
                if val is None:
                    # Handle missing data - for now, fill with mean of genotype or None?
                    # GXE analysis often requires complete data. 
                    # Simple imputation: Genotype Mean
                    g_vals = [data_map.get((g, env)) for env in environments if data_map.get((g, env)) is not None]
                    val = sum(g_vals) / len(g_vals) if g_vals else 0
                row.append(val)
            matrix.append(row)
            
        return {
            "yield_matrix": matrix,
            "genotype_names": genotypes,
            "environment_names": environments
        }

# Singleton instance
_gxe_service: Optional[GxEAnalysisService] = None


def get_gxe_service() -> GxEAnalysisService:
    """Get or create G×E analysis service singleton"""
    global _gxe_service
    if _gxe_service is None:
        _gxe_service = GxEAnalysisService()
    return _gxe_service

File: fortran/src/gxe_analysis.f90
!===============================================================================
! G×E Interaction Analysis Module
! Genotype-by-Environment interaction analysis
!
! Bijmantra Plant Breeding Platform
! Aerospace-Grade Numerical Precision
!
! Methods:
! - AMMI (Additive Main effects and Multiplicative Interaction)
! - GGE Biplot
! - Finlay-Wilkinson Regression
!
! Author: Bijmantra Team
! License: MIT
!===============================================================================

module gxe_analysis
    use iso_c_binding
    use iso_fortran_env, only: real64, int32
    use pca_svd, only: compute_svd
    implicit none
    
    private
    public :: ammi_analysis, gge_biplot, finlay_wilkinson
    
    integer, parameter :: dp = real64
    
contains

    !---------------------------------------------------------------------------
    ! AMMI Analysis (Additive Main effects and Multiplicative Interaction)
    !
    ! Model: Y_ij = μ + G_i + E_j + Σ λ_k α_ik γ_jk + ε_ij
    !
    ! Parameters:
    !   Y          - Yield matrix (g x e), genotypes in rows, environments in cols
    !   g_effects  - Output: Genotype main effects (g)
    !   e_effects  - Output: Environment main effects (e)
    !   g_scores   - Output: Genotype IPCA scores (g x k)
    !   e_scores   - Output: Environment IPCA scores (e x k)
    !   eigenvalues- Output: Eigenvalues (k)
    !   grand_mean - Output: Grand mean
    !   g, e       - Number of genotypes and environments
    !   k          - Number of IPCA axes to retain
    !
    ! Returns: 0 on success
    !---------------------------------------------------------------------------
    function ammi_analysis(Y, g_effects, e_effects, g_scores, e_scores, &
                          eigenvalues, grand_mean, g, e, k) &
        result(status) bind(C, name="ammi_analysis")
        
        integer(c_int), intent(in), value :: g, e, k
        real(c_double), intent(in) :: Y(g, e)
        real(c_double), intent(out) :: g_effects(g)
        real(c_double), intent(out) :: e_effects(e)
        real(c_double), intent(out) :: g_scores(g, k)
        real(c_double), intent(out) :: e_scores(e, k)
        real(c_double), intent(out) :: eigenvalues(k)
        real(c_double), intent(out) :: grand_mean
        integer(c_int) :: status
        
        ! Local variables
        real(dp), allocatable :: GE(:,:), U(:,:), S(:), VT(:,:)
        real(dp) :: row_mean, col_mean
        integer :: i, j, min_ge, info
        
        min_ge = min(g, e)
        allocate(GE(g, e), U(g, min_ge), S(min_ge), VT(min_ge, e))
        
        ! Compute grand mean
        grand_mean = sum(Y) / real(g * e, dp)
        
        ! Compute genotype effects (row means - grand mean)
        do i = 1, g
            row_mean = sum(Y(i, :)) / real(e, dp)
            g_effects(i) = row_mean - grand_mean
        end do
        
        ! Compute environment effects (column means - grand mean)
        do j = 1, e
            col_mean = sum(Y(:, j)) / real(g, dp)
            e_effects(j) = col_mean - grand_mean
        end do
        
        ! Compute GE interaction matrix
        do j = 1, e
            do i = 1, g
                GE(i, j) = Y(i, j) - grand_mean - g_effects(i) - e_effects(j)
            end do
        end do
        
        ! SVD of GE interaction matrix
        status = compute_svd(GE, U, S, VT, g, e)
        if (status /= 0) then
            deallocate(GE, U, S, VT)
            return
        end if
        
        ! Extract IPCA scores
        do i = 1, k
            g_scores(:, i) = U(:, i) * sqrt(S(i))
            e_scores(:, i) = VT(i, :) * sqrt(S(i))
            eigenvalues(i) = S(i)**2
        end do
        
        status = 0
        deallocate(GE, U, S, VT)
        
    end function ammi_analysis

    !---------------------------------------------------------------------------
    ! GGE Biplot Analysis
    !
    ! Model: Y_ij - μ - E_j = Σ λ_k α_ik γ_jk
    ! (Environment-centered, genotype + GE combined)
    !
    ! Parameters:
    !   Y          - Yield matrix (g x e)
    !   g_scores   - Output: Genotype scores (g x k)
    !   e_scores   - Output: Environment scores (e x k)
    !   eigenvalues- Output: Eigenvalues (k)
    !   g, e       - Dimensions
    !   k          - Number of PCs
    !   scaling    - Scaling method (0=symmetric, 1=genotype, 2=environment)
    !
    ! Returns: 0 on success
    !---------------------------------------------------------------------------
    function gge_biplot(Y, g_scores, e_scores, eigenvalues, g, e, k, scaling) &
        result(status) bind(C, name="gge_biplot")
        
        integer(c_int), intent(in), value :: g, e, k, scaling
        real(c_double), intent(in) :: Y(g, e)
        real(c_double), intent(out) :: g_scores(g, k)
        real(c_double), intent(out) :: e_scores(e, k)
        real(c_double), intent(out) :: eigenvalues(k)
        integer(c_int) :: status
        
        ! Local variables
        real(dp), allocatable :: Y_centered(:,:), U(:,:), S(:), VT(:,:)
        real(dp) :: col_mean
        integer :: i, j, min_ge
        real(dp) :: scale_g, scale_e
        
        min_ge = min(g, e)
        allocate(Y_centered(g, e), U(g, min_ge), S(min_ge), VT(min_ge, e))
        
        ! Environment-center the data
        do j = 1, e
            col_mean = sum(Y(:, j)) / real(g, dp)
            Y_centered(:, j) = Y(:, j) - col_mean
        end do
        
        ! SVD
        status = compute_svd(Y_centered, U, S, VT, g, e)
        if (status /= 0) then
            deallocate(Y_centered, U, S, VT)
            return
        end if
        
        ! Apply scaling
        do i = 1, k
            select case (scaling)
            case (0)  ! Symmetric
                scale_g = sqrt(S(i))
                scale_e = sqrt(S(i))
            case (1)  ! Genotype-focused
                scale_g = S(i)
                scale_e = 1.0_dp
            case (2)  ! Environment-focused
                scale_g = 1.0_dp
                scale_e = S(i)
            case default
                scale_g = sqrt(S(i))
                scale_e = sqrt(S(i))
            end select
            
            g_scores(:, i) = U(:, i) * scale_g
            e_scores(:, i) = VT(i, :) * scale_e
            eigenvalues(i) = S(i)**2
        end do
        
        status = 0
        deallocate(Y_centered, U, S, VT)
        
    end function gge_biplot

    !---------------------------------------------------------------------------
    ! Finlay-Wilkinson Regression
    !
    ! Model: Y_ij = μ_i + β_i * E_j + ε_ij
    ! where E_j is the environment mean
    !
    ! Parameters:
    !   Y          - Yield matrix (g x e)
    !   means      - Output: Genotype means (g)
    !   slopes     - Output: Regression slopes (g)
    !   r_squared  - Output: R² values (g)
    !   env_index  - Output: Environment index (e)
    !   g, e       - Dimensions
    !
    ! Returns: 0 on success
    !---------------------------------------------------------------------------
    function finlay_wilkinson(Y, means, slopes, r_squared, env_index, g, e) &
        result(status) bind(C, name="finlay_wilkinson")
        
        integer(c_int), intent(in), value :: g, e
        real(c_double), intent(in) :: Y(g, e)
        real(c_double), intent(out) :: means(g)
        real(c_double), intent(out) :: slopes(g)
        real(c_double), intent(out) :: r_squared(g)
        real(c_double), intent(out) :: env_index(e)
        integer(c_int) :: status
        
        ! Local variables
        real(dp) :: grand_mean, env_mean
        real(dp) :: sum_x, sum_y, sum_xy, sum_xx, sum_yy
        real(dp) :: mean_x, mean_y, ss_x, ss_y, ss_xy
        integer :: i, j
        
        ! Compute grand mean
        grand_mean = sum(Y) / real(g * e, dp)
        
        ! Compute environment index (environment mean - grand mean)
        do j = 1, e
            env_mean = sum(Y(:, j)) / real(g, dp)
            env_index(j) = env_mean - grand_mean
        end do
        
        ! Regression for each genotype
        do i = 1, g
            ! Compute genotype mean
            means(i) = sum(Y(i, :)) / real(e, dp)
            
            ! Simple linear regression: Y_ij = a + b * env_index_j
            sum_x = sum(env_index)
            sum_y = sum(Y(i, :))
            sum_xy = 0.0_dp
            sum_xx = 0.0_dp
            sum_yy = 0.0_dp
            
            do j = 1, e
                sum_xy = sum_xy + env_index(j) * Y(i, j)
                sum_xx = sum_xx + env_index(j) * env_index(j)
                sum_yy = sum_yy + Y(i, j) * Y(i, j)
            end do
            
            mean_x = sum_x / real(e, dp)
            mean_y = sum_y / real(e, dp)
            
            ss_x = sum_xx - real(e, dp) * mean_x * mean_x
            ss_y = sum_yy - real(e, dp) * mean_y * mean_y
            ss_xy = sum_xy - real(e, dp) * mean_x * mean_y
            
            ! Slope
            if (abs(ss_x) > 1.0d-10) then
                slopes(i) = ss_xy / ss_x
            else
                slopes(i) = 1.0_dp
            end if
            
            ! R²
            if (abs(ss_y) > 1.0d-10 .and. abs(ss_x) > 1.0d-10) then
                r_squared(i) = (ss_xy * ss_xy) / (ss_x * ss_y)
            else
                r_squared(i) = 0.0_dp
            end if
        end do
        
        status = 0
        
    end function finlay_wilkinson

end module gxe_analysis
